sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/AggregationIterator.scala

```diff
@@ -157,19 +157,45 @@ abstract class AggregationIterator(
       inputAttributes: Seq[Attribute]): (InternalRow, InternalRow) => Unit = {
     val joinedRow = new JoinedRow
     if (expressions.nonEmpty) {
-      val mergeExpressions = functions.zip(expressions).flatMap {
-        case (ae: DeclarativeAggregate, expression) =>
-          expression.mode match {
-            case Partial | Complete => ae.updateExpressions
-            case PartialMerge | Final => ae.mergeExpressions
+      val mergeExpressions =
+        functions.zip(expressions.map(ae => (ae.mode, ae.isDistinct, ae.filter))).flatMap {
+          case (ae: DeclarativeAggregate, (mode, isDistinct, filter)) =>
+            mode match {
+              case Partial | Complete =>
+                if (filter.isDefined) {
+                  ae.updateExpressions.zip(ae.aggBufferAttributes).map {
+                    case (updateExpr, attr) => If(filter.get, updateExpr, attr)
```

Comment:
this requires that the filter already produce same result for same input.

We should make sure the filter is deterministic during analysis.

Reply:
I'm sorry. I don't understand your means.

Reply:
I checked postgreSQL that support deterministic expression .
```
SELECT
  year,
  count(*) FILTER (WHERE random() * gdp_per_capita >= 40000)
FROM
  countries
GROUP BY
  year;
```

Reply:
We can support it later, by improving the implementation. For now we can't support, otherwise we may partially update the agg buffer and producing undefined result.